### ISSUE TITLE
IP persistence storage seems to not clean up

### ISSUE THREAD

**Issue by rcbarnett** [SpiderLabs/ModSecurity#574]
MODSEC-426: It seem that mod_security is not cleaning up the persistence storage for the IP-storage I am using.

Here is a simple test ruleset I am using (nothing else):

SecRule IP:dos_block "@eq 1" \
        "phase:2,block,msg:'IP address blocked due to high connection rate.', \
        id:2003,status:403,severity:'CRITICAL',tag:'DoS'"
SecRule SCRIPT_FILENAME "@rx /www/content/prod/dl/dl.php" \
        "phase:5,chain,t:none,nolog,pass,id:2001,severity:'INFO',tag:'DoS',setvar:IP.dos_counter=+1,expirevar:IP.dos_counter=60"
SecRule IP:dos_counter "@gt 100" \
        "t:none,setvar:IP.dos_block,setvar:!IP.dos_counter,expirevar:IP.dos_block=60"

The ruleset is working fine. But what I am noticing is, that the /var/tmp/ip.pag file is constantly growing. It is currently at >1GiB
size. When I do a "strings" on that file, I see entries like this:

__expire_KEY
1344192686
794.224.68.113_64d9d88c927c58359a3649f30a00e95070bbb8c0
TIMEOUT
3600
__key
794.224.68.113_64d9d88c927c58359a3649f30a00e95070bbb8c0
__name
CREATE_TIME
1344189037
UPDATE_COUNTER
dos_counter
__expire_dos_counter
1344189097
LAST_UPDATE_TIME
1344189086
94.224.68.113_64d9d88c927c58359a3649f30a00e95070bbb8c0
__expire_KEY
1343169406
868.167.225.250_77150c41806fda817314ba9c40e040c598830d5d
TIMEOUT
3600
__key
868.167.225.250_77150c41806fda817314ba9c40e040c598830d5d
__name
CREATE_TIME
1343165769
UPDATE_COUNTER
dos_counter
__expire_dos_counter
1343165829
LAST_UPDATE_TIME
1343165806
68.167.225.250_77150c41806fda817314ba9c40e040c598830dIR
__expire_KEY
1344005533
786.132.126.93_aa739e3aaaa1fbfc8667bb26120e9930df881a82
TIMEOUT
3600
__key
786.132.126.93_aa739e3aaaa1fbfc8667bb26120e9930df881a82
__name
[...]

Now when I convert the UNIX time strings to UTC, it shows me dates, which are already
expired since weeks. I would asume, that if an entry is expired, that the garbage collection
of mod_sec would remove the entry from the ip.pag file, to keep it small. Or am I wrong?

Any assistance would be highly appreciated.

**Comment by djprmf** [comment]
This is still being worked on?
Almost 3 years now....

**Comment by marcstern** [comment]
Is this specific to Apache?
Is there any plan to replace this mechanism by another one not using APR?
If the answer is "no" to the second questions, we can't let this bug forever! This gives a very bad image of the product.

Does anybody have an idea about the cause fo this bug? Is it so complex to fix?

If it's only a warning that we can ignore, shouldn't we disable it (or log it only once per day/week/month with the number of failures)?

**Comment by zimmerle** [comment]
Hi @djprmf,

The back-end used to save the collections has some limitations that is boynd the way that ModSecurity uses it. As an alternative to this gdbm back-end we implemented the memchace test, which did not became so popular.

Others alternatives were also implemented, as the example of the modsec-sdbm-uitl: 

https://github.com/SpiderLabs/modsec-sdbm-util

This utility can be used to list and cleanup/shrink the sdbm file out-of-band. To circumvent this particular "issue" you may want to reduce the collection timeout:

SecCollectionTimeout -  https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual#seccollectiontimeout

And/or, generally, use the collection with more parsimony. Depending on the amount of usage of the persistent collections functionality, even if there was a cleanup, the amount of I/O in your server will be very high due the way it was implemented.

In ModSecurity version 3 (under development) we tackle the persistent collection in other approach: we let the user select the proper back-end regarding the infrastructure that he/she wants to support. That allow us to support a larger number of users, from a simple setup (with less request/second, lets say...) to a high volume ISP (with very high number of requests).

I am tagging this issue as to be fixed as part of the release of ModSecurity version 3.

@marcstern The problem does not seems to be a implementation problem, rather, the back-end selection. Yes we can re-implement it without using apr or using other back-end such as the memcache implementation. That was the case, but the current target is the ModSecurity version 3 where this feature is actually under development.

**Comment by marcstern** [comment]
The problem does not seems to be a
        implementation problem, rather, the back-end selection. Yes we
        can re-implement it without using apr or using other back-end
        such as the memcache implementation. That was the case, but the
        current target is the ModSecurity version 3 where this feature
        is actually under development.
      Sure, but why does this bug occur? Is it a bug in sdbm or the way
      we use it?
    In the former case, we could check with the developers, in the
      second one we could use it in another way?
    Marc
    
    
    On 27-04-2016 20:23, Felipe Zimmerle
      wrote:
    
    
      
      Hi @djprmf,
      The back-end used to save the collections has some limitations
        that is boynd the way that ModSecurity uses it. As an
        alternative to this gdbm back-end we implemented the memchace
        test, which did not became so popular.
      Others alternatives were also implemented, as the example of
        the modsec-sdbm-uitl: 
      https://github.com/SpiderLabs/modsec-sdbm-util
      This utility can be used to list and cleanup/shrink the sdbm
        file out-of-band. To circumvent this particular "issue" you may
        want to reduce the collection timeout:
      SecCollectionTimeout - https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual#seccollectiontimeout
      And/or, generally, use the collection with more parsimony.
        Depending on the amount of usage of the persistent collections
        functionality, even if there was a cleanup, the amount of I/O in
        your server will be very high due the way it was implemented.
      In ModSecurity version 3 (under development) we tackle the
        persistent collection in other approach: we let the user select
        the proper back-end regarding the infrastructure that he/she
        wants to support. That allow us to support a larger number of
        users, from a simple setup (with less request/second, lets
        say...) to a high volume ISP (with very high number of
        requests).
      I am tagging this issue as to be fixed as part of the release
        of ModSecurity version 3.
      @marcstern The problem does not seems
        to be a implementation problem, rather, the back-end selection.
        Yes we can re-implement it without using apr or using other
        back-end such as the memcache implementation. That was the case,
        but the current target is the ModSecurity version 3 where this
        feature is actually under development.
      —
        You are receiving this because you were mentioned.
        Reply to this email directly or view
          it on GitHub

**Comment by zimmerle** [comment]
Hi @marcstern,

Do you mind to get the text in your comment well formatted? it is important to keep the readability of those issues, it is served as knowledge base to many users. I think github allows you to edit your comment. 

While editing the response it will be good if you can describe the circumstances where you are facing this issue, we may able to help by tuning your configuration a little bit.

**Comment by marcstern** [comment]
I have it on a dozen of machines (the others we suppressed the message).
All Redhat/CentOS 6.x with httpd 2.2 (all stack from platform).
We run modsec-sdbm-util every 15 minutes

**Comment by ossie-git** [comment]
Just ran into this issue in a test machine.  Even after running it with -k, the ip.pag file was still very large:

```
root@httpsbackend:~/modsec-sdbm-util# ls -lsh /var/run/modsecurity/data/*
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.dir
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.pag
 36K -rw-r----- 1 www-data www-data  60K Feb  5 10:09 /var/run/modsecurity/data/ip.dir
7.5M -rw-r----- 1 www-data www-data 472M Feb 10 05:10 /var/run/modsecurity/data/ip.pag
root@httpsbackend:~/modsec-sdbm-util# ./modsec-sdbm-util -k /var/run/modsecurity/data/ip.pag 
Checking target directory: /tmp
Target directory exists.
Opening file: /var/run/modsecurity/data/ip.pag
Database ready to be used.

Total of 3 elements processed.
3 elements removed.
Expired elements: 3, inconsistent items: 0
Fragmentation rate: 100.00% of the database is/was dirty data.
root@httpsbackend:~/modsec-sdbm-util# ls -lsh /var/run/modsecurity/data/*
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.dir
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.pag
 36K -rw-r----- 1 www-data www-data  60K Feb  5 10:09 /var/run/modsecurity/data/ip.dir
7.5M -rw-r----- 1 www-data www-data 472M Feb 11 03:51 /var/run/modsecurity/data/ip.pag
root@httpsbackend:~/modsec-sdbm-util# service apache2 restart
root@httpsbackend:~/modsec-sdbm-util# ls -lsh /var/run/modsecurity/data/*
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.dir
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.pag
 36K -rw-r----- 1 www-data www-data  60K Feb  5 10:09 /var/run/modsecurity/data/ip.dir
7.5M -rw-r----- 1 www-data www-data 472M Feb 11 03:51 /var/run/modsecurity/data/ip.pag
root@httpsbackend:~/modsec-sdbm-util# ./modsec-sdbm-util -k /var/run/modsecurity/data/ip.pag 
Checking target directory: /tmp
Target directory exists.
Opening file: /var/run/modsecurity/data/ip.pag
Database ready to be used.

Total of 0 elements processed.
0 elements removed.
Expired elements: 0, inconsistent items: 0
root@httpsbackend:~/modsec-sdbm-util# ls -lsh /var/run/modsecurity/data/*
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.dir
   0 -rw-r----- 1 www-data www-data    0 Feb  5 10:05 /var/run/modsecurity/data/global.pag
 36K -rw-r----- 1 www-data www-data  60K Feb  5 10:09 /var/run/modsecurity/data/ip.dir
7.5M -rw-r----- 1 www-data www-data 472M Feb 11 03:51 /var/run/modsecurity/data/ip.pag
```

**Comment by Rod-Gomes** [comment]
This ticket is 2013, so far! My God, when will that be solved?

**Comment by bostrt** [comment]
Please see https://github.com/SpiderLabs/ModSecurity/issues/576#issuecomment-279909299 Looks to be same issue and same solution.

**Comment by zimmerle** [comment]
This is fixed by #1274.
